src/main/java/com/amazon/opendistroforelasticsearch/security/OpenDistroSecurityPlugin.java

```diff
@@ -541,53 +542,70 @@ public Weight doCache(Weight weight, QueryCachingPolicy policy) {
-                        if(_user != null && (_user instanceof User)) {
+                public void validateReaderContext(ReaderContext readerContext, TransportRequest transportRequest) {
+                    if (transportRequest instanceof InternalScrollSearchRequest) {
+                        final Object _isLocal = readerContext.getFromContext("_opendistro_security_scroll_auth_local");
+                        final Object _user = readerContext.getFromContext("_opendistro_security_scroll_auth");
+                        if (_user != null && (_user instanceof User)) {
                             final User scrollUser = (User) _user;
                             final User currentUser = threadPool.getThreadContext()
                                     .getTransient(ConfigConstants.OPENDISTRO_SECURITY_USER);
-                            if(!scrollUser.equals(currentUser)) {
-                                auditLog.logMissingPrivileges(SearchScrollAction.NAME, transportRequest, context.getTask());
-                                log.error("Wrong user {} in scroll context, expected {}", scrollUser, currentUser);
-                                throw new ElasticsearchSecurityException("Wrong user in scroll context", RestStatus.FORBIDDEN);
+                            if (!scrollUser.equals(currentUser)) {
+                                auditLog.logMissingPrivileges(SearchScrollAction.NAME, transportRequest, null);
+                                log.error("Wrong user {} in reader context, expected {}", scrollUser, currentUser);
+                                throw new ElasticsearchSecurityException("Wrong user in reader context", RestStatus.FORBIDDEN);
                             }
-                        } else if(_isLocal != Boolean.TRUE) {
-                            auditLog.logMissingPrivileges(SearchScrollAction.NAME, transportRequest, context.getTask());
-                            throw new ElasticsearchSecurityException("No user in scroll context", RestStatus.FORBIDDEN);
+                        } else if (_isLocal != Boolean.TRUE) {
+                            auditLog.logMissingPrivileges(SearchScrollAction.NAME, transportRequest, null);
+                            throw new ElasticsearchSecurityException("No user in reader context", RestStatus.FORBIDDEN);
+                        }
+                    }
+                }
+
+                @Override
+                public void onQueryPhase(SearchContext searchContext, long tookInNanos) {
+                    final Map<String, Set<String>> maskedFieldsMap = (Map<String, Set<String>>) HeaderHelper.deserializeSafeFromHeader(threadPool.getThreadContext(),
+                            ConfigConstants.OPENDISTRO_SECURITY_MASKED_FIELD_HEADER);
+                    final String maskedEval = OpenDistroSecurityUtils.evalMap(maskedFieldsMap, indexModule.getIndex().getName());
```

Comment:
Note: This is relatively hot code, it will be called for each query on each shard. The complexity of `evalMap` grows with the complexity of configurations. For complex configurations, evalMap might thus add some overhead to each query. Thus, it might be a good idea to have some load tests for these changes.

Reply:
It is now evaluated only for term (aggregation) queries.